libs/langchain/ragstack_langchain/graph_store/extractors/html_link_edge_extractor.py

```diff
@@ -99,7 +91,7 @@ def extract_one(
 
         hrefs = _parse_hrefs(input, url, self.drop_fragments)
 
-        link_tags = get_link_tags(document.metadata)
-        link_tags.add(IncomingLinkTag(kind=self._kind, tag=url))
+        link_set = get_link_set(document.metadata)
+        link_set.add(IncomingLinkTag(kind=self._kind, tag=url))
         for url in hrefs:
-            link_tags.add(OutgoingLinkTag(kind=self._kind, tag=url))
+            link_set.add(OutgoingLinkTag(kind=self._kind, tag=url))
```

Comment:
This is unrelated to this PR, but would it be possible to push this logic into the Document? From a usability perspective I feel like the best  API would be something like

```python
document.link_to(kind=..., tag=...)
document.link_from(...)
document.link_bidir(...)
```

I suspect that would require workarounds we don't like (subclassing `Document` or monkey-patching or similar) but maybe it's possible?

Alternatively, I would probably prefer mutating methods, like

```python
from ragstack_knowledge_store.links import link_to, link_from, link_bidir

link_to(document, kind=..., tag=...)
# etc
```

The thing I dislike about the example above is that it doesn't look like the document is actually being modified - `get_link_set` returns a value which is implicitly shared with the document, and that just feels wrong to me. It also introduces the concept of a "link set", which makes it harder to explain what's happening (I say this from experience trying to explain code samples to several people). A casual user shouldn't need to understand how the graph metadata is stored - they just want to create edges. A mutating function is more clear in that it's clear what you're doing (linking things together), what you're operating on (the document) and how you're modifying it (kind, tags, etc).

Reply:
That was introduced in a previous PR, so definitely unrelated to this. I don't think that formulation is ideal since it requires more access via the metadata.

Unfortunately, adding this for document is a bit tricky -- methods on Document would be ugly to add.

I think the best option is:

```
def add_links(doc_or_md: Union["langchain_core.documents.Document", Dict[str, Any]], *links: Link) -> None:
    ...
```

Note it has to do some weird typing, since the `Document` import may not be available -- this may clear up when @cbornet moves this around as part of moving to langchain.

Reply:
👍 `add_links` seems like an improvement.